chasm/test_component_test.go

```diff
@@ -0,0 +1,307 @@
+// THE SOFTWARE.
+
+// TODO: move this to chasm_test package
+package chasm
+
+import (
+	"fmt"
+	"reflect"
+	"sort"
+	"strings"
+
+	commonpb "go.temporal.io/api/common/v1"
+	persistencespb "go.temporal.io/server/api/persistence/v1"
+)
+
+type (
+	// TestComponent is a sample CHASM component used in tests.
+	// It would be nice to move it another package, but this creates a circular dependency.
+
+	protoMessageType = persistencespb.ActivityInfo // Random proto message.
+	TestComponent    struct {
+		UnimplementedComponent
+
+		ComponentData *protoMessageType
+		SubComponent1 Field[*TestSubComponent1]
+		SubComponent2 Field[*TestSubComponent2]
+		SubData1      Field[*protoMessageType]
+	}
+
+	TestSubComponent1 struct {
+		UnimplementedComponent
+
+		SubComponent1Data *protoMessageType
+		SubComponent11    Field[*TestSubComponent11]
+		SubData11         Field[*protoMessageType] // Random proto message.
+	}
+
+	TestSubComponent11 struct {
+		UnimplementedComponent
+
+		SubComponent11Data *protoMessageType
+	}
+
+	TestSubComponent2 struct {
+		UnimplementedComponent
+		SubComponent2Data protoMessageType
+	}
+)
+
+func setTestComponentFields(c *TestComponent) {
+	c.ComponentData = &protoMessageType{
+		ActivityId: "component-data",
+	}
+	c.SubComponent1 = NewComponentField[*TestSubComponent1](nil, &TestSubComponent1{
+		SubComponent1Data: &protoMessageType{
+			ActivityId: "sub-component1-data",
+		},
+		SubComponent11: NewComponentField[*TestSubComponent11](nil, &TestSubComponent11{
+			SubComponent11Data: &protoMessageType{ // Random proto type
+				ActivityId: "sub-component1-sub-component11-data",
+			},
+		}),
+		SubData11: NewDataField[*protoMessageType](nil, &protoMessageType{
+			ActivityId: "sub-component1-sub-data11",
+		}),
+	})
+	c.SubComponent2 = NewEmptyField[*TestSubComponent2]()
+	c.SubData1 = NewDataField[*protoMessageType](nil, &protoMessageType{
+		ActivityId: "sub-data1",
+	})
+}
+
+// returns serialized version of TestComponent from above.
+func testComponentSerializedNodes() map[string]*persistencespb.ChasmNode {
+	serializedNodes := map[string]*persistencespb.ChasmNode{
+		"": &persistencespb.ChasmNode{
+			Metadata: &persistencespb.ChasmNodeMetadata{
+				InitialVersionedTransition: &persistencespb.VersionedTransition{
+					NamespaceFailoverVersion: 1,
+					TransitionCount:          1,
+				},
+				LastUpdateVersionedTransition: &persistencespb.VersionedTransition{
+					NamespaceFailoverVersion: 2,
+					TransitionCount:          2,
+				},
+				Attributes: &persistencespb.ChasmNodeMetadata_ComponentAttributes{
+					ComponentAttributes: &persistencespb.ChasmComponentAttributes{
+						Type:  "test_component",
+						Tasks: []*persistencespb.ChasmComponentAttributes_Task(nil),
+					},
+				},
+			},
+			Data: &commonpb.DataBlob{
+				EncodingType: 1,
+				Data:         []byte{0x42, 0xe, 0x63, 0x6f, 0x6d, 0x70, 0x6f, 0x6e, 0x65, 0x6e, 0x74, 0x2d, 0x64, 0x61, 0x74, 0x61},
```

Comment:
😮 
shall we call proto.Marshal on the data and put results here?

Reply:
It is generated code. Let's use it for some time and see if it is convenient or not.